data_science/scripts/toggle_asg.py

```diff
@@ -3,63 +3,145 @@
 """
 Start/stop all the instances in an autoscaling group.
 
-Usage: toggle_asg.py (--start | --stop | --status) [--type=(p2 | t2)]
+Usage: toggle_asg.py (--start | --stop | --enable-overnight | --disable-overnight) [--type=(p2 | t2)]
+       toggle_asg.py [--type=(p2 | t2)]
+
 
 Actions:
   --start                 Start the autoscaling group (set the desired count to 1).
   --stop                  Stop the autoscaling group (set the desired count to 0).
-  --status                Report the current side of autoscaling group.
+  --enable-overnight      Keeps the autoscaling group desired count at 1 (disables auto scale down)
+  --disable-overnight     Allows the automatic scaling action to be set desired count to 0 (enables auto scale down)
   --type=(p2 | t2)        AWS Instance type (valid values: p2,t2) defaults to t2
 
 """
 
-import sys
-
 import boto3
 import docopt
 
 from asg_utils import discover_asg, set_asg_size
 
+
+def update_desired_capacity_for_scheduled_action(client, scaling_group_name, scheduled_action_name, desired_capacity):
+    """Updates the desired capacity for a scheduled action.
+
+    Keyword arguments:
+    client -- boto3 autoscaling client
+    scaling_group_name -- name of the autoscaling group to modify
+    scheduled_action_name -- name of the scheduled action to modify
+    desired_capacity -- desired capacity of scheduled action
+    """
+
+    action_str = (
+        "Setting desired capacity of ScheduledAction {scheduled_action_name} to {desired_capacity}."
+    ).format(
+        scheduled_action_name=scheduled_action_name,
+        desired_capacity=desired_capacity
+    )
+
+    print(action_str)
+
+    client.put_scheduled_update_group_action(
+        AutoScalingGroupName=scaling_group_name,
+        ScheduledActionName=scheduled_action_name,
+        Recurrence='0 20 * * *',
+        DesiredCapacity=desired_capacity
+    )
+
+
+def get_status(client, tag_name, scheduled_action_name):
+    """Gets the status of an autoscaling group.
+
+    Keyword arguments:
+    client -- boto3 autoscaling client
+    tag_name -- tags used to identify autoscaling group
+    scheduled_action_name -- name of the scheduled action to get status for
+    """
+    scaling_group = discover_asg(asg_client=client, tag_name=tag_name)
+
+    desired_capacity = scaling_group['DesiredCapacity']
+    instance_count = len(scaling_group['Instances'])
+    scaling_group_name = scaling_group['AutoScalingGroupName']
+
+    response = client.describe_scheduled_actions(
+        AutoScalingGroupName=scaling_group_name,
+        ScheduledActionNames=[scheduled_action_name]
+    )
+
+    scheduled_desired_capacity = response['ScheduledUpdateGroupActions'][0]['DesiredCapacity']
```

Comment:
What is ScheduledUpdateGroupActions doesn't have length 1? Maybe add an assert that prints it if not, so we get some debugging info if something goes wrong.